datafusion/core/src/physical_optimizer/sort_pushdown.rs

```diff
@@ -132,6 +172,43 @@ fn pushdown_requirement_to_children(
             RequirementsCompatibility::Compatible(adjusted) => Ok(Some(vec![adjusted])),
             RequirementsCompatibility::NonCompatible => Ok(None),
         }
+    } else if let Some(sort_exec) = plan.as_any().downcast_ref::<SortExec>() {
```

Comment:
I wonder if there is some general way to formulate these conditions about requirements pushdown that doesn't have to special case the operators.

I worry that this logic basically can't be used for user defined execution plan nodes. 

Maybe we could pull the `pushdown_requirement_to_children` function or something into the `ExecutionPlan`

Not needed for this PR, I am just mentioning it

Reply:
I agree, we should find a better approach for handling these cases. We should either introduce new properties for operators , or as you said we should move this logic to methods. Either way we should somehow remove usages of the `.as_any().downcast_ref` usages in the rules.